### PR TITLE
support of noClassPath in getFields

### LINKED ISSUE #775: CtTypeImpl.getAllFields may throw a SpoonClassNotFoundException
I would like to access all fields a of `CtType` node with `getAllFields` (in `NoClasspath` mode). Unfortunately, this method may throw a `SpoonClassNotFoundException` if the accessed class derives from another class not available in classpath. The following snippet show the corresponding stack trace:

```java
spoon.support.reflect.reference.SpoonClassNotFoundException: cannot load class: MimeMessage with class loader sun.misc.Launcher$AppClassLoader@55f96302
	at spoon.support.reflect.reference.CtTypeReferenceImpl.findClass(CtTypeReferenceImpl.java:141)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.getActualClass(CtTypeReferenceImpl.java:129)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.getAllFields(CtTypeReferenceImpl.java:392)
	at spoon.support.reflect.declaration.CtTypeImpl.getAllFields(CtTypeImpl.java:437)
	at spoon.support.reflect.declaration.CtTypeImpl.getAllFields(CtTypeImpl.java:62)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.getAllFields(CtTypeReferenceImpl.java:394)
	at spoon.support.reflect.declaration.CtTypeImpl.getAllFields(CtTypeImpl.java:437)
	at spoon.support.reflect.declaration.CtTypeImpl.getAllFields(CtTypeImpl.java:62)

```

Is this behavior a bug or to be expected?

### PR DESCRIPTION
**Pull request by arturbosch**
…hen the parent class is not found in noClassPath-Mode

This tries to fix #775.

### CODE REVIEW

**Comment by monperrus** [comment]
thanks for this pull request!

this changes the intended fail-fast semantics of `getActualClass`. For #755, I would rather go for an appropriate try/catch in `getAllFields`.

**Comment by arturbosch** [comment]
Something like that?

```java
	@Override
	public Collection<CtFieldReference<?>> getAllFields() {
		CtType<?> t = getDeclaration();
		try {
			if (t == null) {
				return RtHelper.getAllFields(getActualClass(), getFactory());
			} else {
				return t.getAllFields();
			}
		} catch (SpoonClassNotFoundException cnfe) {
			return handleParentNotFound(cnfe);
		}
	}

	private Collection<CtFieldReference<?>> handleParentNotFound(SpoonClassNotFoundException cnfe) {
		String msg = "cannot load class: " + getQualifiedName() + " with class loader "
				+ Thread.currentThread().getContextClassLoader();
		if (getFactory().getEnvironment().getNoClasspath()) {
			// should not be thrown in 'noClasspath' environment (#775)
			Launcher.LOGGER.warn(msg);
			return Collections.emptyList();
		} else {
			throw cnfe;
		}
	}
```

Maybe you have a suggestion for a better message name. Cheers.

**Comment by arturbosch** [comment]
Furthermore getDeclaredFields has the same behaviour and maybe needs a try-catch too.

But now getXXXFields will always return an empty list even if 'Foo' has a field.

```java
public class Foo extends Unknown {
	private String s;
...
}
```

**Comment by monperrus** [comment]
> Something like that?

exactly. should the try/catch be around the RtHelper call directly?

**Comment by retuxx** [comment]
Is it worth merging now?

File: src/test/java/spoon/test/api/NoClasspathTest.java [hunk 1]
```diff
@@ -53,11 +55,17 @@ public void test() throws Exception {
 		}
 		assertNull(superclass.getDeclaration());
 
+		Collection<CtFieldReference<?>> allFields = superclass.getAllFields();
+		System.out.println(allFields);
+		System.out.println(superclass.getDeclaredFields());
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/api/NoClasspathTest.java]: Can you remove these sout?

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java [hunk 2]
```diff
@@ -388,11 +409,15 @@ public void replace(CtTypeReference<?> reference) {
...
-		}
+			if (t == null) {
+				try {
+					return RtHelper.getAllFields(getActualClass(), getFactory());
+				} catch (SpoonClassNotFoundException cnfe) {
+					return handleParentNotFound(cnfe);
+				}
+			} else {
+				return t.getAllFields();
+			}
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java]: Can you remove one level of indentation?

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java [hunk 3]
```diff
@@ -357,24 +359,43 @@ public void replace(CtTypeReference<?> reference) {
...
 
+	private void createFieldReferences(Collection<CtFieldReference<?>> l) {
+		for (Field f : getActualClass().getDeclaredFields()) {
+			l.add(getFactory().Field().createReference(f));
+		}
+		if (getActualClass().isAnnotation()) {
+			for (Method m : getActualClass().getDeclaredMethods()) {
+				CtTypeReference<?> retRef = getFactory().Type().createReference(m.getReturnType());
+				CtFieldReference<?> fr = getFactory().Field().createReference(this, retRef, m.getName());
+				// fr.
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java]: remove comment?

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java [hunk 4]
```diff
@@ -357,24 +359,43 @@ public void replace(CtTypeReference<?> reference) {
...
+				return handleParentNotFound(cnfe);
 			}
-
 		} else {
 			return t.getDeclaredFields();
 		}
 		return l;
 	}
 
+	private void createFieldReferences(Collection<CtFieldReference<?>> l) {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java]: some improvements:
- rename it in `getDeclaredFieldReferences`
- don't take `l` as parameter, create it inside and return it.
- add a one line javadoc

**Comment by arturbosch** [comment]
I incorporated all of your feedback. Sorry, I was in a hurry last time.

**Comment by GerardPaligot** [comment]
Thanks @arturbosch for your last commit but it seems that you removed your call to `superclass.getAllFields();`.

Can you add an assert on this call to show we don't throw anymore an exception? After that, we'll merge your PR.

**Comment by arturbosch** [comment]
Done. Is the comment ok or should I remove it ;) ?

**Comment by GerardPaligot** [comment]
Thanks @arturbosch! :)

**Pull request merged**